src/main/java/org/springframework/data/elasticsearch/core/routing/DefaultRoutingResolver.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core.routing;
+
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentEntity;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentProperty;
+import org.springframework.data.elasticsearch.core.mapping.SimpleElasticsearchMappingContext;
+import org.springframework.data.elasticsearch.core.mapping.SimpleElasticsearchPersistentEntity;
+import org.springframework.data.mapping.PersistentPropertyAccessor;
+import org.springframework.data.mapping.context.MappingContext;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * @author Peter-Josef Meisch
+ */
+public class DefaultRoutingResolver implements RoutingResolver {
+
+	private final SimpleElasticsearchMappingContext mappingContext;
+
+	public DefaultRoutingResolver(MappingContext mappingContext) {
+
+		Assert.isInstanceOf(SimpleElasticsearchMappingContext.class, mappingContext,
+				"mappingContext must be of type SimpleElasticsearchMappingContext");
+
+		this.mappingContext = (SimpleElasticsearchMappingContext) mappingContext;
+	}
+
+	/**
+	 * First tries to resolve the #routing as a property of the bean. If this property exists, it's value is read. If this
+	 * value is not null, to String() is called on the value. If the configured routing is not a property, then a SpEL
+	 * resolution is attempted { @see {@link ElasticsearchPersistentEntity#resolveRouting(Object)}}
+	 */
+	@Override
+	@Nullable
+	public String getRouting(@Nullable Object bean, @Nullable String routing) {
```

Comment:
This method is either called without or with a bean for two well-defined cases. 

I'd suggest to have two distinct methods `getRouting(Class)` and `getRouting(Class, Object)`. Handing in the routing argument leaks the fact that routing can be defined on the type already. Maybe it would even make sense to have `getRouting(IndexCoordinates)` as all these calls to `routingResolver.getRouting(null, null)` are dealing with `null` in the first place.

Reply:
I now have `getRouting()` for the fixed-value case and `getRouting(object bean)`. So it's no longer possible dot use a lambda for `routingResolver` in `operations.withRouting(routingResolver)`, but with `RoutingResolver.just(fixedValue)` this is ok.